### ISSUE TITLE
Add Vulkan/Metal rendering support

### ISSUE THREAD

**Issue by kblaschke** [projectM-visualizer/projectm#683]
Currently, libprojectM only supports rendering using OpenGL 3 and GLES 2. Since OpenGL is generally considered a deprecated API and Vulkan/Metal has a better API to perform multi-threaded rendering and shader compilation, it will be a good step forward to add support for these next-gen rendering APIs.

Any API-specific initialization calls, e.g. providing the required Vulkan structs, should be implemented in a specific header, e.g. `libprojectM/renderer_vulkan.h`. Adding support for multiple rendering backends will also require a major refactoring of the library, abstracting rendering-related calls. projectM only renders a few primitives (triangle fans/strips, line strips and points) and uses vertex and fragment shader programs, so there's not much to implement.

Line rendering should be improved beforehand, as covered by issue #682.

**Comment by olilarkin** [comment]
You could also consider sticking to GLES and using libangle to translate it to Metal (could even be used on windows so it works on top of direct 3D etc)

**Comment by olilarkin** [comment]
Just thought I'd mention it since I've had some recent success with the metal port of angle

**Comment by kblaschke** [comment]
GLES is very limited, and also doesn't allow proper multithreading. Using Vulkan/Metal directly would give some performance improvements, especially when switching presets, as shaders and other resources could be set up in a dedicated parallel thread instead of the render thread, which currently causes hiccups.

**Comment by jherico** [comment]
is any work being done on this currently or is it up for grabs?

**Comment by revmischa** [comment]
> is any work being done on this currently or is it up for grabs?

No work is being done, please feel free!

**Comment by kblaschke** [comment]
> is any work being done on this currently or is it up for grabs?

Anyone who wants to do this, please go ahead!

If you're implementing it, please _keep_ OpenGL/GLES support in projectM. While OpenGL is deprecated, it's still the only way to get projectM running on some devices and systems, one of them being WebGL/Emscripten.

**Comment by tom-leamon** [comment]
> Supporting Metal und Vulkan should be relatively straightforward, as both support GLSL shaders, not requiring any additional shader translation layer. 

That's not true, Metal uses MSL and Vulkan uses SPIR-V. Both require a compilation step from GLSL. For Vulkan you'd use  something like google/shaderc to convert to SPIR-V. sdl3-shadercross can be used to convert from SPIR-V to MSL. That's what I'm doing in my project.

**Comment by mbaetgen-wup** [comment]
I have been doing some testing with ANGLE on Windows. Unfortunately, it does not fully support GLES 3.2 yet. ANGLE is limited to GLES 3.1 with D3D, and GLES 3.0 with Metal. They aim at supporting 3.2 on all platforms at some point, but so far seems just Vulkan supports 3.2. Too bad, as ANGLE is supported by [gl resolver](https://github.com/projectM-visualizer/projectm/pull/942) , it would have been a low hanging fruit. I'll test ANGLE+Vulkan when I get to it.

**Comment by tom-leamon** [comment]
> My main concern here is that we've added numerous quirks to the HLSLParser code to emulate undocumented HLSL behavior many presets rely on to render properly.

> I don't think any proper shader transpiler will "implement" these bugs when transpiling from HLSL, as this is undefined behavior and normally, shader programmers should avoid such things. 

That will definitely be necessary, probably per-target shader language for all the edge cases. A visual regression testing system through image comparisons would be really helpful for this whole process. To solve the problem of presets rendering differently based on timing and audio, you could freeze the time and pass in a static waveform.
